docs/proposals/Certificate Revocation Lists.md

```diff
@@ -0,0 +1,188 @@
+
+**Why Refresh CRL Cache**
+
+A CRL is considered expired if the current date is after the `NextUpdate` field in the CRL. Verify that the CRL is valid (not expired) is necessary for revocation check. Monitoring and refreshing CRL on a regular basis can help avoid CRL download timetaken when doing verification by ensure the CRL is valid.
+
+## Proposed Design
+
+![image](../img/CRL/CRL-workflow.png)
+
+
+Ratify Verification Request Path:
+
+Step 1: Apply the CRs including certs and CRL config
+Step 2: Load CRLs from cert provided URLs // Implement CanCacheCRL() with GetCertificates() which includes all scenarios that introduce a new cert: KMP, KMP refresher and Verifer Config
+Step 3: Trigger Refresh Monitor and set up refresh schedule // Refresher is based on build-in ticker
+Step 4: Start verify task // Revocation list check is handled by notation verifier.
+Step 5: Load trust policy // Get `Opt.Fetcher`
+Step 6: Load CRL cache
+
+CRL Handler:
+
+Step 1: Load cert URLs from `[]*x509.Certificate`
+Step 2: Download CRL
+Step 3: Trigger Refresh Monitor, refresh monitor is [`time`](https://pkg.go.dev/time#example-NewTicker) pkg based.
+
+### Cache Content Design
+
+    
+Key: 
+- `uri` in type `string`
+
+Value: 
+- `*Bundle`
+    - `ttl`, `NextUpdate`
+
+
+Check CRL Cache Validity
+```
+// creates an expiring cache, normal 1 week
+expires := bundle.Metadata.CreatedAt.Add(cache.MaxAge)
+if cache.MaxAge > 0 && time.Now().After(expires) {
+	return nil, ErrCacheMiss
+}
+
+// directly checks CRL validity
+now := time.Now()
+if !crl.NextUpdate.IsZero() && now.After(crl.NextUpdate) {
+	return fmt.Errorf("expired CRL. Current time %v is after CRL NextUpdate %v", now, crl.NextUpdate)
+}
+```
+
+### Load CRL Cache
+
+Load cache is triggerred after cert loaded from the either configurations.
+
+#### Download CRL
+
+
+Download is implemented by CRL `fetcher`
+
+CRL download location (URL) can be obtained from the certificate's CRL Distribution Point (CDP) extension. 
+If the certificate contains multiple CDP locations then each location download is attempted in sequential order, until a 2xx response is received for any of the location. 
+
+For each CDP location, Notary Project verification workflow will try to download the CRL for the default threshold of 5 seconds. The user may be able to configure this threshold. If the CRL cannot be downloaded within the timeout threshold the revocation result will be "revocation unavailable".
+
+#### Save CRL to Cache
+
+```
+// Set stores the CRL bundle in the file system
+// Check closest expired date and set to `CRLCacheProvider`.
+// Save to temp file and avoid concurrency issue with atomic write operation
+// `rename()` is atomic on UNIX-like platforms
+```
+
+### Provide CRL Cache
+
+#### Get Cache from Provider
+
+```
+// Get retrieves the CRL bundle from the file system
+//
+// If the key does not exist, return ErrNotFound
```

Comment:
is this error blocking ( will not continue with validation)? do we think we will see more cases where there will be first attempt failed, subsequent calls succeed. Wondering are we looking to do some perf/stress test here.

Reply:
// Policy that uses custom verification level to relax the strict verification.
            // It logs expiry and skips the revocation check.
            "name": "use-expired-blobs",
            "signatureVerification": {
              "level" : "strict",
              "override" : {
                "expiry" : "log",
                "revocation" : "skip"
              }
            },